### PR TITLE
FancyArrowPatch docstring overhaul

### PR DESCRIPTION
**Pull request by dstansby**
An overhaul of the FancyArrowPatch docstring. There's a couple of kwargs I'm not too sure about so I've left them blank; suggestions welcome.

### CODE REVIEW

File: lib/matplotlib/patches.py [hunk 1]
```diff
@@ -4004,46 +4004,76 @@ def __init__(self, posA=None, posB=None,
...
-        (matplotlib.patches.ConnectionStlye) or a string of the
-        connectionstyle name, with optional comma-separated
-        attributes.  The following connection styles are available.
+        Alternatively if *path* is provided, an arrow is drawn along this path
+        and *patchA*, *patchB*, *shrinkA*, and *shrinkB* are ignored.
 
-        %(AvailableConnectorstyles)s
+        Parameters
+        ----------
+        posA, posB: None, tuple, optional (default: None)
```
Comment (NelleV) [inline lib/matplotlib/patches.py]: Can you add a space before the column? Numpydoc doesn't format properly elsewise.

File: lib/matplotlib/patches.py [hunk 2]
```diff
@@ -4004,46 +4004,76 @@ def __init__(self, posA=None, posB=None,
...
+            (x,y) co-ordinates of arrow tail and arrow head respectively.
```
Comment (NelleV) [inline lib/matplotlib/patches.py]: I think it should be "coordinates".

File: lib/matplotlib/patches.py [hunk 3]
```diff
@@ -4004,46 +4004,76 @@ def __init__(self, posA=None, posB=None,
...
+        arrowstyle: ArrowStyle or string, optional (default: 'simple')
+            Describes how the fancy arrow will be
+            drawn. It can be string of the available arrowstyle names,
+            with optional comma-separated attributes, or an
+            :class:`ArrowStyle` instance. The optional attributes are meant to
+            be scaled with the *mutation_scale*. The following arrow styles are
+            available:
 
-        %(AvailableArrowstyles)s
+            %(AvailableArrowstyles)s
```
Comment (NelleV) [inline lib/matplotlib/patches.py]: We are working towards removing docstring interpolation. If the available arrowstyles are documented else where, can you replace this with the location of the documentation instead of the interpolation? If it isn't documented elsewhere, we can focus on this later.

File: lib/matplotlib/patches.py [hunk 4]
```diff
@@ -3990,10 +3990,10 @@ def __str__(self):
         if self._posA_posB is not None:
             (x1, y1), (x2, y2) = self._posA_posB
             return self.__class__.__name__ \
-                   + "(%g,%g->%g,%g)" % (x1, y1, x2, y2)
```
Comment (anntzer) [inline lib/matplotlib/patches.py]: `"{}({:g},{:g}->{:g},{:g})".format(type(self).__name__, x1, y1, x2, y2)` (or similar with `%` formatting)

File: lib/matplotlib/patches.py (hunk 3 above)
Comment (dstansby) [inline lib/matplotlib/patches.py]: I'm not too sure how to do this - do you know of another PR or commit that has done this replacement?

File: lib/matplotlib/patches.py (hunk 4 above)
Comment (dstansby) [inline lib/matplotlib/patches.py]: I'd rather not touch the actual code in this PR if that's okay?

File: lib/matplotlib/patches.py (hunk 3 above)
Comment (afvincent) [inline lib/matplotlib/patches.py]: I may be wrong but I think @NelleV is referring to this kind of [stuff](http://www.sphinx-doc.org/en/stable/markup/inline.html). Unfortunately, I am not sure that the arrow styles are document elsewhere than in `mpatches.ArrowStyle._style_list `, which has much less information than `_pprint_styles(ArrowStyle._style_list)`.

File: lib/matplotlib/patches.py (hunk 4 above)
Comment (anntzer) [inline lib/matplotlib/patches.py]: sure

File: lib/matplotlib/patches.py [hunk 5]
```diff
@@ -4004,46 +4004,77 @@ def __init__(self, posA=None, posB=None,
...
+                 shrinkB=2,
+                 mutation_scale=1,
                  mutation_aspect=None,
-                 dpi_cor=1.,
+                 dpi_cor=1,
                  **kwargs):
         """
-        If *posA* and *posB* is given, a path connecting two point are
-        created according to the connectionstyle. The path will be
+        If *posA* and *posB* are given, a path connecting two points are
```
Comment (anntzer) [inline lib/matplotlib/patches.py]: ... the two points is created.

File: lib/matplotlib/patches.py [hunk 6]
```diff
@@ -4004,46 +4004,77 @@ def __init__(self, posA=None, posB=None,
...
+        connector:
+
+        patchA, patchB: None, Patch, optional (default: None)
+            Head and tail patch respectively. :class:`matplotlib.patch.Patch`
+            instance.
+
+        shrinkA, shrinkB: scalar, optional (default: 2)
+
+        mutation_scale: scalar, optional (default: 1)
+            Value with which attributes of arrowstyle (e.g., head_length)
```
Comment (anntzer) [inline lib/matplotlib/patches.py]: Add some more `*` here and below?

File: lib/matplotlib/patches.py [hunk 7]
```diff
@@ -4075,28 +4106,40 @@ def __init__(self, posA=None, posB=None,
         self._mutation_aspect = mutation_aspect
 
         self.set_dpi_cor(dpi_cor)
-        #self._draw_in_display_coordinate = True
+        # self._draw_in_display_coordinate = True
```
Comment (anntzer) [inline lib/matplotlib/patches.py]: are there other mentions of this attribute in the codebase?  if it's just a relic of old code it should just be removed.

File: lib/matplotlib/patches.py [hunk 8]
```diff
@@ -4105,65 +4148,82 @@ def set_positions(self, posA, posB):
...
         self.stale = True
 
     def set_connectionstyle(self, connectionstyle, **kw):
         """
-        Set the connection style.
+        Set the connection style. Old attributes are forgotten.
+
+        Parameters
+        ----------
+        connectionstyle: None, ConnectionStyle instance, or string
```
Comment (anntzer) [inline lib/matplotlib/patches.py]: `None, ConnectionStyle, or str`?

File: lib/matplotlib/patches.py [hunk 9]
```diff
@@ -4105,65 +4148,82 @@ def set_positions(self, posA, posB):
...
-        Return the ConnectionStyle instance
+        Return the :class:`ConnectionStyle` instance.
         """
         return self._connector
 
     def set_arrowstyle(self, arrowstyle=None, **kw):
         """
-        Set the arrow style.
+        Set the arrow style. Old attributes are forgotten. Without arguments
+        (or with `arrowstyle=None`) returns available box styles as a list of
```
Comment (anntzer) [inline lib/matplotlib/patches.py]: double backticks

File: lib/matplotlib/patches.py (hunk 6 above)
Comment (NelleV) [inline lib/matplotlib/patches.py]: noooooooooooooooooo :(

File: lib/matplotlib/patches.py (hunk 8 above)
Comment (NelleV) [inline lib/matplotlib/patches.py]: Can you also add a space after the column? It won't render properly elsewise.

Comment (dstansby) [inline lib/matplotlib/patches.py]: @NelleV I'm confused as to what you mean, my local build seems to look fine.

Comment (NelleV) [inline lib/matplotlib/patches.py]: It should be `connectionstyle : None, ConnectionStyle instance, or string`

File: lib/matplotlib/patches.py [hunk 10]
```diff
@@ -4004,46 +4004,77 @@ def __init__(self, posA=None, posB=None,
...
+            available:
 
-        *mutation_scale* : a value with which attributes of arrowstyle
-            (e.g., head_length) will be scaled. default=1.
+            %(AvailableArrowstyles)s
 
-        *mutation_aspect* : The height of the rectangle will be
-            squeezed by this value before the mutation and the mutated
-            box will be stretched by the inverse of it. default=None.
+        arrow_transmuter :
```
Comment (QuLogic) [inline lib/matplotlib/patches.py]: This does not appear to be used; maybe it should not be documented? Or marked deprecated or something.

File: lib/matplotlib/patches.py [hunk 11]
```diff
@@ -4004,46 +4004,77 @@ def __init__(self, posA=None, posB=None,
...
+        connectionstyle : str, ConnectionStyle, or None, optional
+        (default: 'arc3')
+            Describes how *posA* and *posB* are connected. It can be an
+            instance of the :class:`ConnectionStyle` class or a string of the
+            connectionstyle name, with optional comma-separated attributes. The
+            following connection styles are available:
+
+            %(AvailableConnectorstyles)s
 
+        connector :
```
Comment (QuLogic) [inline lib/matplotlib/patches.py]: Also seems to be unused. Actually, both of these are used in `Connector`, which subclasses from `FancyArrowPatch`, so I think maybe someone added it here to give it the same API. Maybe just mark these two as ignored.

File: lib/matplotlib/patches.py [hunk 12]
```diff
@@ -4004,46 +4004,77 @@ def __init__(self, posA=None, posB=None,
...
+
+        patchA, patchB : None, Patch, optional (default: None)
+            Head and tail patch respectively. :class:`matplotlib.patch.Patch`
+            instance.
+
+        shrinkA, shrinkB : scalar, optional (default: 2)
```
Comment (QuLogic) [inline lib/matplotlib/patches.py]: "Shrinking factor of tail and head of arrow"?

File: lib/matplotlib/patches.py [hunk 13]
```diff
@@ -4105,65 +4148,83 @@ def set_positions(self, posA, posB):
...
     def set_connectionstyle(self, connectionstyle, **kw):
         """
-        Set the connection style.
+        Set the connection style. Old attributes are forgotten.
+
+        Parameters
+        ----------
+        connectionstyle : None, ConnectionStyle instance, or string
+            Can be a string with connectionstyle name with
+            optional comma-separated attributes, eg::
```
Comment (QuLogic) [inline lib/matplotlib/patches.py]: e.g.

File: lib/matplotlib/patches.py [hunk 14]
```diff
@@ -4105,65 +4148,83 @@ def set_positions(self, posA, posB):
...
 
-        *connectionstyle* can be a string with connectionstyle name with
-         optional comma-separated attributes. Alternatively, the attrs can be
-         provided as keywords.
+                set_connectionstyle("arc,angleA=0,armA=30,rad=10")
 
-         set_connectionstyle("arc,angleA=0,armA=30,rad=10")
-         set_connectionstyle("arc", angleA=0,armA=30,rad=10)
+            Alternatively, the attributes can be provided as keywords, eg::
```
Comment (QuLogic) [inline lib/matplotlib/patches.py]: e.g.

File: lib/matplotlib/patches.py [hunk 15]
```diff
@@ -4105,65 +4148,83 @@ def set_positions(self, posA, posB):
...
+        strings.
 
-        *arrowstyle* can be a string with arrowstyle name with optional
-         comma-separated attributes. Alternatively, the attrs can
-         be provided as keywords.
+        Parameters
+        ----------
+        arrowstyle : None, ArrowStyle, str, optional (default: None)
+            Can be a string with arrowstyle name with optional comma-separated
+            attributes, eg.::
```
Comment (QuLogic) [inline lib/matplotlib/patches.py]: e.g.

File: lib/matplotlib/patches.py [hunk 16]
```diff
@@ -4105,65 +4148,83 @@ def set_positions(self, posA, posB):
...
 
-         set_arrowstyle("Fancy,head_length=0.2")
-         set_arrowstyle("fancy", head_length=0.2)
+                set_arrowstyle("Fancy,head_length=0.2")
 
-        Old attrs simply are forgotten.
+            Alternatively attributes can be provided as keywords, eg.::
```
Comment (QuLogic) [inline lib/matplotlib/patches.py]: e.g.

File: lib/matplotlib/patches.py [hunk 17]
```diff
@@ -4177,30 +4238,38 @@ def set_arrowstyle(self, arrowstyle=None, **kw):
...
         """
         self._mutation_scale = scale
         self.stale = True
 
     def get_mutation_scale(self):
         """
         Return the mutation scale.
+
+        Returns
+        ----------
```
Comment (QuLogic) [inline lib/matplotlib/patches.py]: Extra dashes.

**Comment by anntzer** [comment]
@dstansby did react to @QuLogic's comments so let's assume until further notice that he plans to address the issues.

**Comment by QuLogic** [comment]
They've been handled below the lines I commented on; GitHub is not hiding the diff though.

**Comment by anntzer** [comment]
Ah, I see.
